### PR TITLE
Improve `ShaderLab` compilation error log and package build

### PR DESCRIPTION
**Pull request by Sway007**
### Please check if the PR fulfills these requirements

- [x] The commit message follows our [guidelines](https://github.com/galacean/engine/blob/main/.github/COMMIT_MESSAGE_CONVENTION.md)
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

#### change rollup config to produce 2 version `ShaderLab` package dist: `Release` and `Editor`.

```ts
// use release version
import { ShaderLab } from "@galacean/engine-shader-lab"

// use editor version

// umd
import { ShaderLab } from "@galacean/engine-shader-lab/dist/browser.editor";
// esmoudle
import { ShaderLab } from "@galacean/engine-shader-lab/dist/module.editor";
// commonjs
import { ShaderLab } from "@galacean/engine-shader-lab/dist/main.editor";
```

#### Improve the error log when compiation error of ShaderLab occur in `Editor` version.

```
6 errors occur!
ShaderLab.ts:167
CompilationError: Invalid engine constant: BlendFactor.SourceAlph22

|···                BlendState {
|···                  Enabled = true;
|···                  SourceColorBlendFactor = BlendFactor.SourceAlph22;
                                                           ^^^^^^^^^^^^
|···                  DestinationColorBlendFactor = BlendFactor.OneMinusSourceAlpha;
|···                  SourceAlphaBlendFactor = BlendFactor.One;

CompilationError: Invalid render queue Transparent22

|···                  WriteEnabled = false;
|···                }
|···                RenderQueueType = Transparent22;
                                      ^^^^^^^^^^^^^
|···              #else
|···                BlendState {

CompilationError: Invalid "BlendState" variable: blendState

|···              #endif
|···
|···              BlendState = blendState;
                               ^^^^^^^^^^
|···
|···              mat4 renderer_MVPMat;

CompilationError: undeclared identifier:attr2

|···                Varyings v;
|···              
|···                gl_Position = renderer_MVPMat * attr2.POSITION;
                                                    ^^^^^
|···                none();
|···                v.v_pos = gl_Position.xyz;

CompilationError: No overload function type found: none

|···              
|···                gl_Position = renderer_MVPMat * attr2.POSITION;
|···                none();
                    ^^^^^^
|···                v.v_pos = gl_Position.xyz;
|···                v.v_uv = attr.TEXCOORD_023;

CompilationError: referenced attribute not found: TEXCOORD_023

|···                none();
|···                v.v_pos = gl_Position.xyz;
|···                v.v_uv = attr.TEXCOORD_023;
                                  ^^^^^^^^^^^^
|···                return v;
|···              }
```

> In `Release` version, `ShaderLab` will immediately terminate the compilation and throw when encountering an error.

related issue: #2248 

## Summary by CodeRabbit

- **New Features**
	- Enhanced error handling in shader processing, providing detailed error messages and context for debugging.
	- Introduced new error types for better categorization of issues, including `PreprocessorError`, `CompilationError`, and `ScannerError`.
	- Added a new `verbose` version of ShaderLab for improved diagnostic information during development.
	- Integrated a new method for scanning characters in the shader source.
	- Introduced a utility class for managing object pools and error handling in shader lab elements.

- **Chores**
	- Streamlined the build process by removing unnecessary editor-related scripts from the configuration.
	- Standardized the build command across CI workflows for consistency.
	- Updated test cases to improve coverage and added new tests for handling compilation errors.

### CODE REVIEW

**Comment by Sway007** [comment]
> 建议使用 LogLevel 的设计去实现编译细节差异，而不是针对平台去设计不同的实现，平台是内部的期望，不具有通用性意义，且维护成本大，容易烂尾。有机会的话建议把以前的 build:editor 都改成通用的设计。

条件编译最大的优势是减小最终打包的包体尺寸，运行时还能去除if else比较的开销。缺点是维护性欠佳，主要是因为js语言没有原生的宏编译，只能借助rollup插件实现。现在release版本包体尺寸比editor版本包体尺寸少30kb (109kb / 139kb)

File: packages/shader-lab/src/Error.ts [hunk 1]
```diff
@@ -1,19 +1,76 @@
...
+    let diagnosticMessage = `${this.name}: ${message}\n\n`;
+    const lineSplit = "|···";
+
+    for (let i = 0; i <= end.line + GSError.wrappingLineCount; i++) {
+      if (i < start.line - GSError.wrappingLineCount) continue;
+
+      diagnosticMessage += lineSplit;
+      diagnosticMessage += `${lines[i]}\n`;
+      if (i >= start.line && i <= end.line) {
+        diagnosticMessage += " ".repeat(lineSplit.length + start.column);
```
[the remark this answers is not in the document]
Comment (Sway007) [reply packages/shader-lab/src/Error.ts:46]: 这个类都是在Editor版本里的，release版本没有这些性能损耗

File: packages/shader-lab/src/Error.ts [hunk 2]
```diff
@@ -1,19 +1,76 @@
...
+        diagnosticMessage += "^".repeat(Math.max(end.column - start.column, 1));
```
[the remark this answers is not in the document]
Comment (Sway007) [reply packages/shader-lab/src/Error.ts:47]: 同上

File: packages/shader-lab/src/Error.ts [hunk 3]
```diff
@@ -1,19 +1,76 @@
...
+  }
+
+  log(_source?: string): void {
+    if (!Logger.enable) return;
+    let start: ShaderPosition, end: ShaderPosition;
+    const { message, loc, source: originSource } = this;
+    let source = originSource;
+    if (_source) source = _source;
+    if (!source) {
+      Logger.error(message);
```
[the remark this answers is not in the document]
Comment (Sway007) [reply packages/shader-lab/src/Error.ts:26]: done

File: packages/shader-lab/src/Error.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (zhuxudong) [reply packages/shader-lab/src/Error.ts:47]: 哥，Editor 就不用管性能了吗

File: packages/shader-lab/src/ShaderLab.ts [hunk 4]
```diff
@@ -47,6 +35,17 @@ export class ShaderLab implements IShaderLab {
     return range;
   }
 
+  // #if _EDITOR
+  private _errors: GSError[] = [];
+
+  /**
+   * Retrieve the compilation errors
+   */
+  get errors(): GSError[] | undefined {
```
Comment (zhuxudong) [inline packages/shader-lab/src/ShaderLab.ts:44]: 不是最多空数组么，怎么会有 undefined

File: packages/shader-lab/src/contentParser/Scanner.ts [hunk 5]
```diff
@@ -25,4 +25,11 @@ export default class Scanner extends BaseScanner {
     }
     return Number(this._source.substring(start, this._currentIndex));
   }
+
+  scanToCharacter(char: string): void {
+    while (this.getCurChar() !== char) {
```
Comment (zhuxudong) [inline packages/shader-lab/src/contentParser/Scanner.ts:30]: 这里用户不小心写错，不会死循环吗

Comment (Sway007) [reply packages/shader-lab/src/contentParser/Scanner.ts]: done

File: packages/shader-lab/src/ShaderLab.ts (hunk 4 above)
Comment (Sway007) [reply packages/shader-lab/src/ShaderLab.ts]: done

File: packages/shader-lab/src/Error.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (Sway007) [reply packages/shader-lab/src/Error.ts]: done

File: packages/shader-lab/src/Error.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (Sway007) [reply packages/shader-lab/src/Error.ts]: done

File: packages/core/src/SystemInfo.ts [hunk 6]
```diff
@@ -8,6 +8,10 @@ export class SystemInfo {
   static platform: Platform = Platform.Unknown;
   /** The operating system is running on. */
   static operatingSystem: string = "";
+
+  /** @internal */
+  static _isBrowser = true;
```
Comment (GuoLei1990) [inline packages/core/src/SystemInfo.ts:13]: Merge latest dev/1.4

File: packages/shader-lab/README.md [hunk 7]
```diff
@@ -26,6 +26,19 @@ const shader = Shader.create(galaceanShaderCode);
 engine.run()
 ```
 
+There are two versions of ShaderLab: `Release` and `Debug`. The Debug version offers more user-friendly diagnostic information for debug ShaderLab compilation errors, while the Release version provides superior performance.
+
+you can use debug version by import:
+
+```ts
+// umd
+import { ShaderLab } from "@galacean/engine-shader-lab/dist/browser.editor";
```
Comment (GuoLei1990) [inline packages/shader-lab/README.md:35]: `browser.editor` `Release` `Debug`，概念混乱

Comment (Sway007) [reply packages/shader-lab/README.md]: done

File: packages/core/src/SystemInfo.ts (hunk 6 above)
Comment (Sway007) [reply packages/core/src/SystemInfo.ts]: done

File: packages/shader-lab/package.json [hunk 8]
```diff
@@ -25,7 +25,8 @@
   },
   "files": [
     "dist/**/*",
-    "types/**/*"
+    "types/**/*",
+    "verbose"
```
Comment (GuoLei1990) [inline packages/shader-lab/package.json:29]: `verbose` mean folder?

File: packages/shader-lab/src/Error.ts [hunk 9]
```diff
@@ -1,19 +1,79 @@
...
+    Logger.error(diagnosticMessage);
+  }
+}
+
+export class PreprocessorError extends GSError {
+  constructor(message: string, loc: ShaderRange | ShaderPosition, source: string, file?: string, cause?: Error) {
+    super(message, loc, source, file, cause);
+    this.name = "PreprocessorError";
+  }
+}
```
Comment (GuoLei1990) [inline packages/shader-lab/src/Error.ts:64]: Tell me what is the value of using abstract classes？

File: packages/shader-lab/src/Error.ts [hunk 10]
```diff
@@ -1,19 +1,79 @@
...
+  readonly file?: string;
+
+  constructor(message: string, loc: ShaderRange | ShaderPosition, source: string, file?: string, cause?: Error) {
     super(message, { cause });
     this.loc = loc;
+    this.source = source;
+    this.file = file;
+  }
+
+  log(_source?: string): void {
```
Comment (GuoLei1990) [inline packages/shader-lab/src/Error.ts:19]: fix param name, check all!

File: packages/shader-lab/src/Error.ts [hunk 11]
```diff
@@ -1,19 +1,79 @@
...
+      if (i < start.line - GSError.wrappingLineCount) continue;
+
+      diagnosticMessage += lineSplit;
+      diagnosticMessage += `${lines[i]}\n`;
+      const paddingLength = lineSplit.length + start.column;
+      const remarkLength = Math.max(end.column - start.column, 1);
+      if (i >= start.line && i <= end.line) {
+        diagnosticMessage += " ".repeat(paddingLength);
+        diagnosticMessage += "^".repeat(remarkLength);
+        diagnosticMessage += "\n";
```
Comment (GuoLei1990) [inline packages/shader-lab/src/Error.ts:51]: Waste performance

File: packages/shader-lab/src/ShaderLab.ts [hunk 12]
```diff
@@ -2,42 +2,30 @@ import { Lexer } from "./lexer";
...
+  static createPosition(index: number, line?: number, column?: number): ShaderPosition {
     const position = this._shaderPositionPool.get();
-    position.setX(
-      index,
-      // #if _EDITOR
-      line,
-      column
-      // #endif
-    );
+    position.setX(index, line, column);
```
Comment (GuoLei1990) [inline packages/shader-lab/src/ShaderLab.ts:28]: setX?

Comment (Sway007) [reply packages/shader-lab/src/ShaderLab.ts]: change to `set`

File: packages/shader-lab/src/Error.ts (hunk 11 above)
Comment (Sway007) [reply packages/shader-lab/src/Error.ts]: done

File: packages/shader-lab/src/Error.ts (hunk 10 above)
Comment (Sway007) [reply packages/shader-lab/src/Error.ts]: done

File: packages/shader-lab/package.json (hunk 8 above)
Comment (Sway007) [reply packages/shader-lab/package.json]: done

File: packages/shader-lab/src/Error.ts (hunk 9 above)
Comment (Sway007) [reply packages/shader-lab/src/Error.ts]: done

File: packages/shader-lab/src/Error.ts [hunk 13]
```diff
@@ -1,19 +1,79 @@
 // #if _EDITOR
```
[the remark this answers is not in the document]
Comment (Sway007) [reply packages/shader-lab/src/Error.ts]: all replaced by _VERBOSE

File: packages/shader-lab/src/Error.ts [hunk 14]
```diff
@@ -1,19 +1,72 @@
-// #if _EDITOR
-import { ShaderRange } from "./common";
+// #if _VERBOSE
+import { Logger } from "@galacean/engine";
+import { ShaderPosition, ShaderRange } from "./common";
 
-export abstract class GSError extends Error {
-  readonly loc: ShaderRange;
+export class GSError extends Error {
```
Comment (GuoLei1990) [inline packages/shader-lab/src/Error.ts:5]: File name use main class name: `Error.ts`->`GSError.ts`

File: packages/shader-lab/src/Error.ts [hunk 15]
```diff
@@ -1,19 +1,72 @@
-// #if _EDITOR
-import { ShaderRange } from "./common";
+// #if _VERBOSE
+import { Logger } from "@galacean/engine";
+import { ShaderPosition, ShaderRange } from "./common";
```
Comment (GuoLei1990) [inline packages/shader-lab/src/Error.ts:3]: Include file path is better

File: packages/shader-lab/src/Error.ts (hunk 14 above)
Comment (Sway007) [reply packages/shader-lab/src/Error.ts]: done

File: packages/shader-lab/src/ShaderLab.ts [hunk 16]
```diff
@@ -2,42 +2,30 @@ import { Lexer } from "./lexer";
...
 // @ts-ignore
 import { Logger, ShaderLib, ShaderMacro, ShaderPass, ShaderPlatformTarget } from "@galacean/engine";
 import { ShaderPosition, ShaderRange } from "./common";
 import { ShaderLabObjectPool } from "./ShaderLabObjectPool";
+// #if _EDITOR
+import { GSError } from "./Error";
+// #endif
+import { PpParser } from "./preprocessor/PpParser";
 
 export class ShaderLab implements IShaderLab {
```
[the remark this answers is not in the document]
Comment (Sway007) [reply packages/shader-lab/src/ShaderLab.ts:17]: done

File: packages/shader-lab/src/Error.ts (hunk 15 above)
Comment (Sway007) [reply packages/shader-lab/src/Error.ts]: done

File: packages/shader-lab/src/preprocessor/MacroDefine.ts [hunk 17]
```diff
@@ -1,6 +1,6 @@
 import { BaseToken } from "../common/BaseToken";
 import { ShaderRange } from "../common";
-import { ParserUtils } from "../Utils";
+import { GSError, GSErrorName } from "../GSError";
 
 export class MacroDefine {
```
Comment (GuoLei1990) [inline packages/shader-lab/src/preprocessor/MacroDefine.ts:5]: Use `public readonly macro: BaseToken` to opt in constructor

Comment (GuoLei1990) [inline packages/shader-lab/src/preprocessor/MacroDefine.ts:5]: `expand` is `expandFunction`?

File: packages/shader-lab/src/preprocessor/PpParser.ts [hunk 18]
```diff
@@ -662,25 +693,25 @@ export default class PpParser {
         const range = ShaderLab.createRange(token.location!.start, scanner.curPosition);
         replace = macro.expand(...args);
```
Comment (GuoLei1990) [inline packages/shader-lab/src/preprocessor/PpParser.ts:694]: Waste performance
```suggestion
        replace = macro.expand(...args);
```

File: packages/shader-lab/src/ShaderLab.ts [hunk 19]
```diff
@@ -2,51 +2,48 @@ import { Lexer } from "./lexer";
...
 
   static createRange(start: ShaderPosition, end: ShaderPosition): ShaderRange {
     const range = this._shaderRangePool.get();
-    range.setX(start, end);
+    range.set(start, end);
     return range;
   }
 
+  // #if _VERBOSE
+  private _errors: GSError[] = [];
```
Comment (GuoLei1990) [inline packages/shader-lab/src/ShaderLab.ts:37]: public readonly

File: packages/shader-lab/src/ShaderLab.ts [hunk 20]
```diff
@@ -66,68 +62,80 @@ export class ShaderLab implements IShaderLab {
...
 
     Logger.info(`[pass compilation - preprocessor]  cost time ${performance.now() - preprocessorStart}ms`);
 
     const lexer = new Lexer(ppdContent);
     const tokens = lexer.tokenize();
-    const program = ShaderLab._parser.parse(tokens);
+
+    const { _parser: parser } = ShaderLab;
+
+    ShaderLab._processingPassText = ppdContent;
```
Comment (GuoLei1990) [inline packages/shader-lab/src/ShaderLab.ts:84]: `_processingPassText` do what?

File: packages/shader-lab/src/ShaderLab.ts [hunk 21]
```diff
@@ -66,68 +62,80 @@ export class ShaderLab implements IShaderLab {
...
 
   _parseShaderContent(shaderSource: string): IShaderContent {
+    ShaderLabUtils.clearAllShaderLabObjectPool();
     ShaderContentParser.reset();
-    return ShaderContentParser.parse(shaderSource);
+    // #if _VERBOSE
+    this._errors.length = 0;
+    // #endif
+    const ret = ShaderContentParser.parse(shaderSource);
+    for (const error of ShaderContentParser._errors) {
```
Comment (GuoLei1990) [inline packages/shader-lab/src/ShaderLab.ts:122]: Forget to add `_VERBOSE`?

File: packages/shader-lab/src/codeGen/VisitorContext.ts [hunk 22]
```diff
@@ -52,26 +58,44 @@ export class VisitorContext {
...
-      return;
+      // #if _VERBOSE
+      return new GSError(
+        GSErrorName.CompilationError,
+        `referenced varying not found: ${ident.lexeme}`,
+        ident.location,
+        ShaderLab._processingPassText
+      );
+      // #else
+      // @ts-ignore
```
Comment (GuoLei1990) [inline packages/shader-lab/src/codeGen/VisitorContext.ts:94]: delete `@ts-ignore`

File: packages/shader-lab/src/common/BaseScanner.ts [hunk 23]
```diff
@@ -35,13 +39,15 @@ export default class BaseScanner {
...
 
   get curPosition(): ShaderPosition {
-    return ShaderLab.createPosition(
-      this._currentIndex,
-      // #if _EDITOR
-      this._column,
-      this._line
-      // #endif
-    );
+    return ShaderLab.createPosition(this._currentIndex, this._line, this._column);
```
Comment (GuoLei1990) [inline packages/shader-lab/src/common/BaseScanner.ts:42]: Why delete `// #if _VERBOSE`?

Comment (GuoLei1990) [inline packages/shader-lab/src/common/BaseScanner.ts:42]: @image

`get curPosition` ->`createPosition()` is better!
```ts
 get curPosition(): ShaderPosition {
    return ShaderLab.createPosition(this._currentIndex, this._line, this._column);
  }
```

File: packages/shader-lab/src/common/BaseScanner.ts [hunk 24]
```diff
@@ -65,24 +71,18 @@ export default class BaseScanner {
...
-   */
   _advance(): void {
     if (this.isEnd()) {
       return;
     }
 
-    this._currentIndex++;
-
-    // #if _EDITOR
     if (this.getCurChar() === "\n") {
```
Comment (GuoLei1990) [inline packages/shader-lab/src/common/BaseScanner.ts:79]: Why delete _VERBOSE

File: packages/shader-lab/src/common/BaseScanner.ts [hunk 25]
```diff
@@ -124,11 +124,21 @@ export default class BaseScanner {
     this.skipCommentsAndSpace();
     const peek = this.peek(text.length);
     if (peek !== text) {
-      ParserUtils.throw(this._currentIndex, `Expect ${text}, got ${peek}`);
+      this.throwError(this.curPosition, `Expect text "${text}", but got "${peek}"`);
     }
     this.advance(text.length);
   }
 
+  throwError(pos: ShaderPosition | ShaderRange, ...msgs: any[]) {
```
Comment (GuoLei1990) [inline packages/shader-lab/src/common/BaseScanner.ts:132]: reportError? throwError?

File: packages/shader-lab/src/contentParser/Scanner.ts [hunk 26]
```diff
@@ -25,4 +25,11 @@ export default class Scanner extends BaseScanner {
     }
     return Number(this._source.substring(start, this._currentIndex));
   }
+
+  scanToCharacter(char: string): void {
```
Comment (GuoLei1990) [inline packages/shader-lab/src/contentParser/Scanner.ts:29]: Forget to add _VERBOSE

Comment (Sway007) [reply packages/shader-lab/src/contentParser/Scanner.ts:30]: done

File: packages/shader-lab/src/common/BaseScanner.ts (hunk 24 above)
Comment (Sway007) [reply packages/shader-lab/src/common/BaseScanner.ts:89]: Done, all column and line properties are move to verbose mode.

File: packages/shader-lab/src/common/BaseScanner.ts (hunk 23 above)
Comment (Sway007) [reply packages/shader-lab/src/common/BaseScanner.ts]: done

Comment (Sway007) [reply packages/shader-lab/src/common/BaseScanner.ts]: done

File: packages/shader-lab/src/common/BaseScanner.ts [hunk 27]
```diff
@@ -98,20 +110,20 @@
...
     } else if (this.peek(2) === "/*") {
-      const start = this.curPosition;
+      const start = this.getCurPosition();
       this.advance(2);
       //  multi-line comments
       while (this.peek(2) !== "*/" && !this.isEnd()) this._advance();
       this.advance(2);
       this.skipCommentsAndSpace();
-      return ShaderLab.createRange(start, this.curPosition);
+      return ShaderLab.createRange(start, this.getCurPosition());
```
[the remark this answers is not in the document]
Comment (Sway007) [reply packages/shader-lab/src/common/BaseScanner.ts:126]: yes, show me some unit tests for multi-line comment parsing

File: packages/shader-lab/src/codeGen/VisitorContext.ts (hunk 22 above)
Comment (Sway007) [reply packages/shader-lab/src/codeGen/VisitorContext.ts]: done

File: packages/shader-lab/src/codeGen/GLESVisitor.ts [hunk 28]
```diff
@@ -26,6 +29,7 @@ export abstract class GLESVisitor extends CodeGenVisitor {
   abstract getVaryingDeclare(): ICodeSegment[];
 
   visitShaderProgram(node: ASTNode.GLShaderProgram, vertexEntry: string, fragmentEntry: string): IShaderInfo {
+    this._errors.length = 0;
```
[the remark this answers is not in the document]
Comment (Sway007) [reply packages/shader-lab/src/codeGen/GLESVisitor.ts]: done

File: packages/shader-lab/src/codeGen/CodeGenVisitor.ts [hunk 29]
```diff
@@ -1,17 +1,26 @@
...
 
 /**
+ * @internal
  * The code generator
  */
 export class CodeGenVisitor {
-  protected constructor() {}
+  protected _errors: GSError[] = [];
+
+  get errors() {
```
[the remark this answers is not in the document]
Comment (Sway007) [reply packages/shader-lab/src/codeGen/CodeGenVisitor.ts]: done

File: packages/shader-lab/src/ShaderLab.ts (hunk 19 above)
Comment (Sway007) [reply packages/shader-lab/src/ShaderLab.ts]: done

File: packages/shader-lab/src/preprocessor/PpParser.ts (hunk 18 above)
Comment (Sway007) [reply packages/shader-lab/src/preprocessor/PpParser.ts]: done

File: packages/shader-lab/src/preprocessor/MacroDefine.ts (hunk 17 above)
Comment (Sway007) [reply packages/shader-lab/src/preprocessor/MacroDefine.ts:5]: done

File: packages/shader-lab/src/GSError.ts [hunk 30]
```diff
@@ -0,0 +1,65 @@
...
+    public readonly file?: string
+  ) {
+    super(message);
+    this.name = name;
+  }
+
+  override toString(): string {
+    let start: ShaderPosition, end: ShaderPosition;
+    const { message, location: loc, source } = this;
+    if (!source) {
```
[the remark this answers is not in the document]
Comment (Sway007) [reply packages/shader-lab/src/GSError.ts:22]: done

File: packages/shader-lab/src/ShaderLabUtils.ts [hunk 31]
```diff
@@ -0,0 +1,17 @@
...
+  private static _shaderLabObjectPoolSet: ClearableObjectPool<IPoolElement>[] = [];
+
+  static createObjectPool<T extends IPoolElement>(type: new () => T) {
+    const pool = new ClearableObjectPool<T>(type);
+    ShaderLabUtils._shaderLabObjectPoolSet.push(pool);
+    return pool;
+  }
+
+  static clearAllShaderLabObjectPool() {
+    for (let i = 0; i < ShaderLabUtils._shaderLabObjectPoolSet.length; i++) {
```
[the remark this answers is not in the document]
Comment (Sway007) [reply packages/shader-lab/src/ShaderLabUtils.ts:16]: done

File: packages/shader-lab/src/common/BaseScanner.ts (hunk 27 above)
[the remark this answers is not in the document]
Comment (Sway007) [reply packages/shader-lab/src/common/BaseScanner.ts:126]: Impressive! Suggestion for test case accepted! Thanks

File: packages/shader-lab/src/GSError.ts [hunk 32]
```diff
@@ -0,0 +1,65 @@
...
+      start = location.start;
+      end = location.end;
+    }
+    const lines = source.split("\n");
+
+    let diagnosticMessage = `${this.name}: ${message}\n\n`;
+    const lineSplit = "|···";
+
+    for (let i = start.line - GSError.wrappingLineCount, n = end.line + GSError.wrappingLineCount; i <= n; i++) {
+      diagnosticMessage += lineSplit + `${lines[i]}\n`;
```
Comment (GuoLei1990) [inline packages/shader-lab/src/GSError.ts:38]: cache `lines[i]`

**Review by GuoLei1990 — APPROVED** [review]

**Pull request merged**